content/blog/2019-01-07-we-mastered-the-art-of-programming.md

```diff
@@ -0,0 +1,63 @@
+
+## Why isn't this working ?
+
+This called for my problem solving skills. How can I achieve faster production? By better understanding what he was doing.
+
+First thing I saw is the pre-processing step was nightly done by a robot. The second thing I saw was his cache, where all the data were ready for treatment and intermediate values were even pre-calculated (for data grouping and other analysis). The third thing was the modeling part, where everything was so tightly bundled that he could multi-thread all different data cross-sections in one operation. The best part of all this is — all those things were done by 9 am when he arrived. Least to say that gave him plenty of time to work on his analysis.
+
+I asked him if he could lend me a piece or two to integrate them into my scripts. As a good teammate, he obliged and provided me with a chunk of code. My pride was hurt enough that I went back to my desk alone and worked on it by myself. It was a relatively big chunk of code, so I expected a proportionate effort to plug everything and test. I opened the code and could quickly locate the line that would require modification. The music of typing was flowing into my ears — yes I had my mechanical keyboard at that time. It only took me around half an hour to do everything. I made it work without any unexpected side effects the first time I launched the script.
+
+What a great modification done by a great programmer.
+
+## Wait.
+
+This is the moment the epiphany came. I was not the one have been doing a great job: my teammate was. He handed me some code that was so clean that anyone could quickly modify anything without unexpected side effect. Everything was self-explanatory, no comments lying around, no arbitrary numbers. The code had the perfect vertical and horizontal length and my solarized theme made the code look like a work of art. I experienced, what I thought at that time was, the holy grail of good code.
+
+## Different solutions.
+
+First of all, there were things my teammate knew that i didn't know that i didn't know. Things like data pipeline, [design patterns](https://sourcemaking.com/design_patterns) and [principles](https://en.wikipedia.org/wiki/SOLID), unit testing, [clean code](https://en.wikipedia.org/wiki/Worship), software architecture, etc. These elements are the normal distribution for the actuary, the mean and variance for the statistician. The ground on which everything is build. Clearly my hacking and school background never prepared me for this.
+
+My teammate had an holistic view of the problem. I was too busy *making things work* while my teammate was busy figuring *the best way* to make it work. This subtle difference is crucial. A good solution will make something work right now. The best solution makes something work right now AND consider the future user. Even if it means taking some extra time to step back and rebuild in the right direction.
```

Comment:
build in the right direction (tu rebuild pas si tu sais où tu vas)

Reply:
xie xie fixed